Fix tooltip error and add fixed steps to linear scale

roben opened the pull request:
I noticed that using the label callbacks only added all returned contents to a single line in contrast to the documentation: "All functions must return either a string or an array of strings. Arrays of strings are treated as multiple lines of text.". This pull request fixes this.

roben commented:
And sorry for the merge conflict, tried to amend after pushing.

roben commented:
I have updated the pull request which now also sports a ticks.fixedStepSize option to force fixed linear scale steps instead of automatically generated ones. In my case I had to display ranking table changes over time and there were positions skipped or even half positions displayed depending on the ranking size / max value. I also fixed a typo from the broken commit earlier.

etimberg commented:
@roben did the tabs / spacing change in the linear scale? It looks like the entire file changed :| It also looks like a really old version of the file. Can you also merge the latest v2.0-dev into your version.

It would also be great if you could add documentation for the new option to the markdown files in the docs folder.

roben commented:
It was the most recent file from v2.0-dev but I used Netbeans' auto format which had different indentation settings than your default. I have changed and reindented it, the diff is much lighter now.
I have added the configuration documentation in a new table for the linear scale section as the parameter only applies there.

I have also added the built files to the commit just for completeness.

Note that I needed to change `"gulp": "3.5.x",` to `"gulp": "3.9.x",` in package.json to prevent gulp 3.5.6 throwing an "unsupported version" error. I did not commit that change, though, as it seems to work for your environment.

etimberg commented on nbproject/private/private.properties:
```diff
@@ -0,0 +1,2 @@
+auxiliary.org-netbeans-modules-javascript-karma.config=karma.conf.js
+browser=Chrome.INTEGRATED
```
Can remove this file and others in the nbproject directory

etimberg commented on src/scales/scale.linear.js:
```diff
@@ -169,16 +170,24 @@
-			var niceMax = Math.ceil(this.max / spacing) * spacing;
+			if (this.options.ticks.fixedStepSize && this.options.ticks.fixedStepSize > 0) {
+				var j = this.min;
+				while (j <= this.max) {
```
It is probably safer to exit the loop when you've added enough spaces rather than when j > max. The reason is that for very small numbers this may fail.

What about setting `numSpaces = this.options.ticks.fixedStepSize` and then setting `spacing = (this.max - this.min) / numSpaces`. That way you can reuse the loop that's in the else block for both cases.

etimberg commented:
@roben thanks for updated. I added a couple of comments to the diff. Once those are addressed, I'll merge this :smile:

thowmas commented:
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  thnghm@gmail.com
    Domain typeshape.com has exceeded the max emails per hour (1/1 (100%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from github-smtp2-ext1.iad.github.net ([192.30.252.192]:35389 helo=github-smtp2a-ext-cp1-prd.iad.github.net)
	by whm01.mammoth.net.au with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.86)
	(envelope-from <noreply@github.com>)
	id 1aRg8Q-000CBJ-7T
	for tom@typeshape.com; Sat, 06 Feb 2016 00:09:25 +1100
Date: Fri, 05 Feb 2016 05:08:30 -0800
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1454677710;
	bh=G4P+CVT2L0zUnI8Ae7yi84W8xKlNb/rJO3IxGg/RZb8=;
	h=From:Reply-To:To:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=mXg58Z8oKWU6Derva9U4Q917ASU7cqHi0lsmOk+nAsMAeQ/yDFIsdBpxSZn+9xWsK
	 grp6JbAZvC2oZMtUK/LOnlcJ3EBw7iD0ghDz0VK9JVmrL+6Q+J0q1XRJLb7tTUUTKF
	 /4H+gvUS/yGm2V3+4QUHO3hjH1DMPCUHaR5Z2+Fw=
From: Evert Timberg <notifications@github.com>
Reply-To: "nnnick/Chart.js" <reply@reply.github.com>
To: "nnnick/Chart.js" <Chart.js@noreply.github.com>
Message-ID: <nnnick/Chart.js/pull/1978/c180349504@github.com>
In-Reply-To: <nnnick/Chart.js/pull/1978@github.com>
References: <nnnick/Chart.js/pull/1978@github.com>
Subject: Re: [Chart.js] V2.0 dev (#1978)
Mime-Version: 1.0
Content-Type: multipart/alternative;
 boundary="--==_mimepart_56b49ece97664_328a3fcc5023d2bc23114ea";
 charset=UTF-8
Content-Transfer-Encoding: 7bit
Precedence: list
X-GitHub-Sender: etimberg
X-GitHub-Recipient: thowmas
List-ID: nnnick/Chart.js <Chart.js.nnnick.github.com>
List-Archive: https://github.com/nnnick/Chart.js
List-Post: <mailto:reply@reply.github.com>
List-Unsubscribe: <mailto:unsub+000ccf01aa0861aa035748e7743809da2f93b705c7070b9e92cf0000000112cc60ce92a169ce07d35f31@reply.github.com>,
 <https://github.com/notifications/unsubscribe/AAzPAbW1DdQm4ll4pxKrrhpaHwOpIQISks5phJZOgaJpZM4HTNtM>
X-Auto-Response-Suppress: All
X-GitHub-Recipient-Address: tom@typeshape.com
X-Spam-Status: No, score=-5.0
X-Spam-Score: -49
X-Spam-Bar: -----
X-Ham-Report: Spam detection software, running on the system "whm01.mammoth.net.au",
 has NOT identified this incoming email as spam.  The original
 message has been attached to this so you can view it or label
 similar future email.  If you have any questions, see
 root\@localhost for details.
 
 Content preview:  @roben thanks for updated. I added a couple of comments to
    the diff. Once those are addressed, I'll merge this :smile: --- Reply to
   this email directly or view it on GitHub: https://github.com/nnnick/Chart.js/pull/1978#issuecomment-180349504
    [...] 
 
 Content analysis details:   (-5.0 points, 5.0 required)
 
  pts rule name              description
 ---- ---------------------- --------------------------------------------------
 -5.0 RCVD_IN_DNSWL_HI       RBL: Sender listed at http://www.dnswl.org/, high
                             trust
                             [192.30.252.192 listed in list.dnswl.org]
 -0.0 RCVD_IN_MSPIKE_H4      RBL: Very Good reputation (+4)
                             [192.30.252.192 listed in wl.mailspike.net]
 -0.0 SPF_PASS               SPF: sender matches SPF record
 -1.9 BAYES_00               BODY: Bayes spam probability is 0 to 1%
                             [score: 0.0000]
  2.1 HTML_IMAGE_ONLY_12     BODY: HTML: images with 800-1200 bytes of words
  0.0 HTML_MESSAGE           BODY: HTML included in message
 -0.1 DKIM_VALID_AU          Message has a valid DKIM or DK signature from author's
                             domain
  0.1 DKIM_SIGNED            Message has a DKIM or DK signature, not necessarily valid
 -0.1 DKIM_VALID             Message has at least one valid DKIM or DK signature
 -0.0 RCVD_IN_MSPIKE_WL      Mailspike good senders
X-Spam-Flag: NO


----==_mimepart_56b49ece97664_328a3fcc5023d2bc23114ea
Content-Type: text/plain;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

@roben thanks for updated. I added a couple of comments to the diff. Once those are addressed, I'll merge this :smile: 

---
Reply to this email directly or view it on GitHub:
https://github.com/nnnick/Chart.js/pull/1978#issuecomment-180349504
----==_mimepart_56b49ece97664_328a3fcc5023d2bc23114ea
Content-Type: text/html;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

<p><a href="https://github.com/roben" class="user-mention">@roben</a> thanks for updated. I added a couple of comments to the diff. Once those are addressed, I'll merge this <img class="emoji" title=":smile:" alt=":smile:" src="https://assets-cdn.github.com/images/icons/emoji/unicode/1f604.png" height="20" width="20" align="absmiddle"> </p>

<p style="font-size:small;-webkit-text-size-adjust:none;color:#666;">&mdash;<br>Reply to this email directly or <a href="https://github.com/nnnick/Chart.js/pull/1978#issuecomment-180349504">view it on GitHub</a>.<img alt="" height="1" src="https://github.com/notifications/beacon/AAzPAcdb4i1QSqOvXwTQ3JKMMRr9eLvhks5phJZOgaJpZM4HTNtM.gif" width="1" /></p>
<div itemscope itemtype="http://schema.org/EmailMessage">
<div itemprop="action" itemscope itemtype="http://schema.org/ViewAction">
  <link itemprop="url" href="https://github.com/nnnick/Chart.js/pull/1978#issuecomment-180349504"></link>
  <meta itemprop="name" content="View Pull Request"></meta>
</div>
<meta itemprop="description" content="View this Pull Request on GitHub"></meta>
</div>

----==_mimepart_56b49ece97664_328a3fcc5023d2bc23114ea--

roben commented:
The nbproject folder should't have been committed at all. I'll remove it.

Regarding the loop: Before implementing my changes I had issues with the graph ignoring suggestedMin and starting with 0.5 instead of 1 and I was confused by all the rounding going on there. I exit the loop after max has been reached to make sure that the last value has a tick label.

But what do you mean with 
> The reason is that for very small numbers this may fail.
? I can't see how this loop could fail besides looking really ugly if the step width is set too small.

etimberg commented:
The case I was thinking of is when `fixedStepSize` is a small number between 0 and 1. It may be that the exact number is not fully representable as a floating point number. It could be that when generating the last tick, the loop bails out one iteration early. #1498 was an earlier bug where I ended up having to change the loop that generates ticks.

roben commented:
I have applied the changes and merged your repository to fix the conflicts.

etimberg commented:
Thanks @roben
